SICP_book/chapter_2_2_metta.metta

```diff
@@ -0,0 +1,1729 @@
+
+; (square-list (list 1 2 3 4)) ; -> (1 4 9 16)
+
+; Here are two different definitions of square-list. Complete both of them by filling in the missing expressions:
+
+(= (square-list Nil)
+    Nil)
+(= (square-list (Cons $x $xs))
+    (Cons (* $x $x) (square-list $xs)))
+
+(= (square-list-m $items)
+    (map square $items))
+
+!(assertEqual
+    (square-list (list (1 2 3 4 5)))
+    (list (1 4 9 16 25)))
+
+!(assertEqual
+    (square-list-m (list (1 2 3 4 5)))
+    (list (1 4 9 16 25)))
+
+; -----------------------End of Exercise 2.21---------------------------
+
+
+; Exercise 2.23.
+; The procedure for-each is similar to map. It takes as arguments a procedure and a list of
+; elements. However, rather than forming a list of the results, for-each just applies the
+; procedure to each of the elements in turn, from left to right. The values returned by applying
+; the procedure to the elements are not used at all -- for-each is used with
+; procedures that perform an action, such as printing. For example,
+
+; (for-each (lambda (x) (newline) (display x))
+;           (list 57 321 88))
+; 57
+; 321
+; 88
+
+; The value returned by the call to for-each (not illustrated above) can
+; be something arbitrary, such as true. Give an implementation of for-each.
+
+; one way to implement for-each is:
+;(= (for-each $proc Nil)
+;    Nil)
+;(= (for-each $proc (Cons $x $xs))
+;    (let*
+;    (
+;        (() ($proc $x))
+;        (() (for-each $proc $xs))
+;    )
+;    ()))
+
+; This one was provided by Vitaly Bogdanov. And this implementation checks if there was an error during $proc call.
+; This is helpful in exercise 2.49 when I'm using for-each to call python function.
+
+(= (for-each $proc Nil)
+    Nil)
+(= (for-each $proc (Cons $x $xs))
+    (let $res ($proc $x)
+            (unify $res (Error $_atom $_msg)
+            $res
+            (for-each $proc $xs))))
+
+; prints list
+!(assertEqual
+    (for-each println! (list (57 321 88)))
+    Nil)
+; -----------------------End of Exercise 2.23---------------------------
+
+; Procedure to count leaves on Scheme looks like this:
+;(define (count-leaves x)
+;    (cond ((null? x) 0)
+;        ((not (pair? x)) 1)
+;        (else (+ (count-leaves (car x))
+;            (count-leaves (cdr x))))))
+
+; But in Metta, we do not have function pair? So, I'm using a workaround - checking type of current input. List in Metta
+; will be something like that (Number Number (Number Number)) for list (1 5 (6 4)). So, when we are reaching "not-pair?"
+; we are actually reaching just Number in our program. Of course this check won't work if we are working with list of
+; letters not numbers, but this is a workaround I've got in mind. Feel free to share your vision.
+
+(= (count-leaves Nil)
+    0)
+(= (count-leaves $x)
+        (if (== (get-type $x) Number)
+            1
+            (empty)))
+(= (count-leaves (Cons $x $xs))
+        (+ (count-leaves $x)
+            (count-leaves $xs)))
```

Comment:
The following implementation doesn't rely on type of the items in the tree:
```
(= (count-leaves $tree)
   (case $tree (
     (Nil 0)
     ((Cons $x $xs) (+ (count-leaves $x) (count-leaves $xs)))
     ($_ 1) )))
```
I would suggest replacing all functions where `(== (get-type $x) Number)` is used by the similar implementation pattern.